bootloaders/riotboot_serial/main.c

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright (C) 2017 Kaspar Schleiser <kaspar@schleiser.de>
+ *                    Inria
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     bootloaders
+ * @{
+ *
+ * @file
+ * @brief      Serial UART riot-based bootloader
+ *
+ * @author     Kaspar Schleiser <kaspar@schleiser.de>
+ * @author     Francisco Acosta <francisco.acosta@inria.fr>
+ * @author     Benjamin Valentin <benjamin.valentin@ml-pa.com>
+ *
+ * @}
+ */
+
+#include "cpu.h"
+#include "panic.h"
+#include "riotboot/slot.h"
+#include "riotboot/serial.h"
+
+static int _get_slot(void)
+{
+    uint32_t version = 0;
+    int slot = -1;
+
+    if (!IS_USED(MODULE_RIOTBOOT_SLOT)) {
+        return -1;
+    }
+
+    for (unsigned i = 0; i < riotboot_slot_numof; i++) {
+        const riotboot_hdr_t *riot_hdr = riotboot_slot_get_hdr(i);
+        if (riotboot_slot_validate(i)) {
+            /* skip slot if metadata broken */
+            continue;
+        }
+        if (riot_hdr->start_addr != riotboot_slot_get_image_startaddr(i)) {
+            continue;
+        }
+        if (slot == -1 || riot_hdr->version > version) {
+            version = riot_hdr->version;
+            slot = i;
+        }
+    }
```

Comment:
From the PR description, I don't quite understand if `riotboot_slot` is supported or not?

Reply:
It's supported but there is no build system integration.

Reply:
What would be required?

Reply:
The easiest would be merging the `main.c` files of the three riotboot implementations and then just select the appropriate module(s).

 The build system hard-codes `bootloaders/riotboot` when using slots.

Reply:
Would this do or I'm still missing stuff?

```
ifneq (,$(filter riotboot_serial,$(USEMODULE)))
  RIOTBOOT_DIR ?= $(RIOTBASE)/bootloaders/riotboot_serial
else ifneq (,$(filter riotboot_dfu,$(USEMODULE)))
  RIOTBOOT_DIR ?= $(RIOTBASE)/bootloaders/riotboot_dfu
else
  RIOTBOOT_DIR ?= $(RIOTBASE)/bootloaders/riotboot
endif